doc: explicitly doc package.exports is breaking (#33074)
nodejs/node: doc/api/esm.md

entry point of the package.

The `"exports"` field provides an alternative to `"main"` where the package
main entry point can be defined while also encapsulating the package,
**preventing any other entry points besides those defined in `"exports"`**.
This encapsulation allows module authors to define a public interface for
their package 

If both `"exports"` and `"main"` are defined, the `"exports"` field takes
precedence over `"main"`. It is important to note that `"exports"` are not
specific to ES modules or CommonJS; `"main"` will be overridden by `"exports"`

> if it exists. As such `"main"` cannot be used as a fallback for `"Common.js"`

Reviewer: ```suggestion
if it exists. As such `"main"` cannot be used as a fallback for CommonJS
```

> but it can be used as a fallback for legacy versions of Node.js that do not
> support ES Modules.

Reviewer: ```suggestion
but it can be used as a fallback for legacy versions of Node.js that do not
support the exports field.
```

[Conditional Exports][] can be used within `"exports"` to define different
package entry points per environment, including whether the package is
referenced via `require` or via `import`. For more information about supporting
both Common.js and ES Modules in a single package please consult
[the dual CommonJS/ES module packages section][]. 

**Warning**: Introducing the `"exports"` field prevents consumers of a package
from using any entry points that are not defined, including the `package.json`.
This will likely be a breaking change. To make the introduction of `"exports"`
non-breaking, ensure that every previously supported entry point is exported.
It is best to explicitly specify entry points so that the package's public API
is well-defined. For example, a project that previous exported `main`, `lib`,

> `feature`, and the `package.json` could use the following `package.exports`

Reviewer: ```suggestion
`feature`, and the `package.json` could use the following `package.exports`:
```

```json
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/index.js": "./lib/index.js",
    "./feature": "./feature/index.js",
    "./feature/index.js": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

> As a last resort, package encapsulation can be disabled entirely by creating an
> export for the root of the package `"./": "./"`. This will expose every file in
> the package at the cost of disabling encapsulation.

Reviewer: ```suggestion
As a last resort, package encapsulation can be disabled entirely by creating an
export for the root of the package `"./": "./"`. This will expose every file in
the package at the cost of disabling encapsulation and the tooling and optimization
benefits that come with it.
```

entry point of the package.

The `"exports"` field provides an alternative to `"main"` where the package
main entry point can be defined while also encapsulating the package,
**preventing any other entry points besides those defined in `"exports"`**.
This encapsulation allows module authors to define a public interface for

> their package 

Reviewer 2: ```suggestion
their package.
```

If both `"exports"` and `"main"` are defined, the `"exports"` field takes
precedence over `"main"`. It is important to note that `"exports"` are not
specific to ES modules or CommonJS; `"main"` will be overridden by `"exports"`
if it exists. As such `"main"` cannot be used as a fallback for CommonJS
but it can be used as a fallback for legacy versions of Node.js that do not
support the exports field.

[Conditional Exports][] can be used within `"exports"` to define different
package entry points per environment, including whether the package is
referenced via `require` or via `import`. For more information about supporting
both Common.js and ES Modules in a single package please consult
[the dual CommonJS/ES module packages section][]. 

> **Warning**: Introducing the `"exports"` field prevents consumers of a package
> from using any entry points that are not defined, including the `package.json`.
> This will likely be a breaking change. To make the introduction of `"exports"`
> non-breaking, ensure that every previously supported entry point is exported.
> It is best to explicitly specify entry points so that the package's public API
> is well-defined. For example, a project that previous exported `main`, `lib`,

Reviewer 2: ```suggestion
**Warning**: Introducing the `"exports"` field prevents consumers of a package
from using any entry points that are not defined, including the `package.json`
(e.g. `require('your-package/package.json')`. **This will likely be a breaking
change.**

To make the introduction of `"exports"` non-breaking, ensure that every
previously supported entry point is exported. It is best to explicitly specify
entry points so that the package’s public API is well-defined. For example,
a project that previous exported `main`, `lib`,
```

`feature`, and the `package.json` could use the following `package.exports`:

```json
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/index.js": "./lib/index.js",
    "./feature": "./feature/index.js",
    "./feature/index.js": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

As a last resort, package encapsulation can be disabled entirely by creating an
export for the root of the package `"./": "./"`. This will expose every file in
the package at the cost of disabling encapsulation and the tooling and optimization

> benefits that come with it.

Reviewer 2: Since these tooling and optimization benefits don't yet exist, it feels a bit premature to point to them as a reason to use `"exports"`. This paragraph comes off a bit as "we at Node want you to use `"exports"`, and to use encapsulation, but we aren't going to really explain why."

There is a straightforward use case for `"./": "./"` (versus not adding `"exports"` at all): a package author who doesn't want encapsulation but who wants to use conditional exports to define separate entry points for CommonJS and ESM. One could argue that `"./": "./"` is the safer choice as it's for-sure not a breaking change, versus spelling out every export and possibly missing one.

Author: I added this phrasing at the request of @guybedford would you like to suggest an alternative?

Author: I've removed the copy about tooling

Reviewer: > Since these tooling and optimization benefits don't yet exist

All my own tooling very much relies on and uses exports for automatic optimization, including the next version of jspm. And I always opt-out of all of these optimizations when `./`: `./` is set.

I would have preferred we included this line as when giving advice it is always helpful to explain the reason for the advice.

That said I understand there has been a silly amount of back and fourth there so I won't press this.

entry point of the package.

The `"exports"` field provides an alternative to `"main"` where the package
main entry point can be defined while also encapsulating the package,
**preventing any other entry points besides those defined in `"exports"`**.
This encapsulation allows module authors to define a public interface for
their package.

If both `"exports"` and `"main"` are defined, the `"exports"` field takes
precedence over `"main"`. It is important to note that `"exports"` are not
specific to ES modules or CommonJS; `"main"` will be overridden by `"exports"`
if it exists. As such `"main"` cannot be used as a fallback for CommonJS
but it can be used as a fallback for legacy versions of Node.js that do not

> support the exports field.

Reviewer 3: ```suggestion
support the `"exports"` field.
```

[Conditional Exports][] can be used within `"exports"` to define different
package entry points per environment, including whether the package is
referenced via `require` or via `import`. For more information about supporting

> both Common.js and ES Modules in a single package please consult

Reviewer 3: ```suggestion
both CommonJS and ES Modules in a single package please consult
```

[the dual CommonJS/ES module packages section][]. 

**Warning**: Introducing the `"exports"` field prevents consumers of a package
from using any entry points that are not defined, including the `package.json`
(e.g. `require('your-package/package.json')`. **This will likely be a breaking
change.**

To make the introduction of `"exports"` non-breaking, ensure that every
previously supported entry point is exported. It is best to explicitly specify
entry points so that the package’s public API is well-defined. For example,
a project that previous exported `main`, `lib`,
`feature`, and the `package.json` could use the following `package.exports`:

```json
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",

>     "./lib/index.js": "./lib/index.js",

Reviewer 3: Do we also need to add `"./lib/index": "./lib/index.js"`?

Author: Done.

from using any entry points that are not defined, including the `package.json`
(e.g. `require('your-package/package.json')`. **This will likely be a breaking
change.**

To make the introduction of `"exports"` non-breaking, ensure that every
previously supported entry point is exported. It is best to explicitly specify
entry points so that the package’s public API is well-defined. For example,
a project that previous exported `main`, `lib`,
`feature`, and the `package.json` could use the following `package.exports`:

```json
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/index": "./lib/index.js",
    "./lib/index.js": "./lib/index.js",
    "./feature": "./feature/index.js",
    "./feature/index.js": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

Alternatively a project could choose to export entire folders:

```json
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/": "./lib/",
    "./feature": "./feature/index.js",
    "./feature/": "./feature/",
    "./package.json": "./package.json"
  }
}
```

As a last resort, package encapsulation can be disabled entirely by creating an

> export for the root of the package `"./": "./"`. This will expose every file in

Reviewer 4: This currently makes it sound like `./: ./` exposes more paths than the individual paths above. Afaict that isn't true since `./: ./` won't expose `./feature` while the above does. This paragraph may make people think that `./: ./` is both shorter to write and will expose more than the previous examples which is a dangerous misunderstanding (it will expose less).

Author: Added some copy PTAL

Reviewer 4: SGTM

change.**

To make the introduction of `"exports"` non-breaking, ensure that every
previously supported entry point is exported. It is best to explicitly specify
entry points so that the package’s public API is well-defined. For example,
a project that previous exported `main`, `lib`,
`feature`, and the `package.json` could use the following `package.exports`:

```json
{
  "name": "my-mod",
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/index": "./lib/index.js",
    "./lib/index.js": "./lib/index.js",
    "./feature": "./feature/index.js",
    "./feature/index.js": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

Alternatively a project could choose to export entire folders:

```json
{
  "name": "my-mod",
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/": "./lib/",
    "./feature": "./feature/index.js",
    "./feature/": "./feature/",
    "./package.json": "./package.json"
  }
}
```

As a last resort, package encapsulation can be disabled entirely by creating an
export for the root of the package `"./": "./"`. This will expose every file in

> the package at the cost of disabling encapsulation. As the ES Module loader in
> Node.js will not resolve file extensions or an `index.js` exporting the root is
> less expressive than either of the above examples as individuals will be unable
> to `import feature from 'my-mod/feature'`.

Reviewer: ```suggestion
the package at the cost of disabling the encapsulation and potential tooling
benefits this provides. As the ES Module loader in Node.js will not resolve file
extensions or an `index.js` exporting the root is less expressive than either of
the above examples as individuals will sill be unable to
`import feature from 'my-mod/feature'`.
```

environment, including whether the package is referenced via `require` or via
`import`.

> ***Warning***: Introducing the exports field prevents consumers of a package

Reviewer 2: ```suggestion
**Warning**: Introducing the `"exports"` field prevents consumers of a package
```

> for using any entry points that is not defined, including the `package.json`.

Reviewer 5: ```suggestion
from using any entry points that are not defined, including the `package.json`.
```

> This is a Semver-Major change. To make the introduction of `"exports"`

Reviewer 6: Not everybody would agree that it's a semver-major change. For my packages, I do not consider deep imports to be part of the public API. Otherwise that would mean that internal refactoring (moving a file, or an internal function from one file to another for example) should also be considered semver-major.

Reviewer 7: Yes, as it should.

The “exports” field changes the discussion from “how strictly do each of us interpret the semver spec” to “there’s no way you could consume something that i didn’t intend”. In the most liberal interpretation of semver, it might not be a breaking change, but its better to tell these people that it is, than to tell the rest of the spectrum that it’s not.

Reviewer 6: I don't think that we should tell that it's not semver-major. But we can have this warning in the documentation (I'm in favor) without explicitly mentioning semverness. I would prefer something like: "This is a breaking change. To prevent the introduction of `"exports"` from affecting current consumers of a package, either ensure ..."

Reviewer 7: What’s the benefit of not mentioning it? The worst case for your constituency is “an unnecessary major”, the worst case for mine is “breaking changes in a nonmajor” - i think we should be erring on the side of unnecessary majors here/

Reviewer 8: ```suggestion
This may be a Semver-Major change. To make the introduction of `"exports"`
```

I would encourage softening this a little bit, by at least allowing for the possibility that a package's API is something different from all-possible-entry-points. As far as I can tell, this would be the only place in the Node.js documentation where an opinion is espressed about the versioning of packages other than Node.js itself.

If the Node.js project does have a strong opinion on the versioning of JS packages, that opinion should probably be communicated more clearly and more strongly than in an aside like this. To be clear, I at least would almost certainly agree with it; I just find it surprising to encounter only in a place like this.

Author: The use of `Semver-Major` and `Semver-Minor` are no longer included in this doc change. I'm going to resolve this conversation PTAL

> This is a Semver-Major change. To make the introduction of `"exports"`

Reviewer 2: ```suggestion
This will likely be a breaking change. To make the introduction of `"exports"`
```

> Semver-Minor either ensure that every previously supported entry point is

Reviewer 2: ```suggestion
non-breaking, either ensure that every previously supported entry point is
```

> exported or include a export of the root of the package `"./"; "./"`.

Reviewer 5: ```suggestion
exported or include an export of the root of the package `"./": "./"`.
```

> exported or include an export of the root of the package `"./": "./"`.

Reviewer: I don't think we should be advising this pattern - if users do this it really defeats the point of exports in not providing any encapsulation.

Any user who is slightly unsure of what to do will read this then just always add the base map - and all the nice static analyzability benefits of packages go back out the window.

I would really much rather users simply don't ship exports than ship it badly.

Reviewer 2: Well I guess the only reason to ship exports with a full map is to provide conditional exports for the main entry point? So maybe the example should reflect that?

Reviewer: The thing is - the correct thing to do is just take the time to full enumerate every subpath that might be required by users. This way we avoid exposing eg test folders and build scripts.

Reviewer 9: Test folders and build scripts should be ignored via the [package `files` field](https://docs.npmjs.com/files/package.json#files), or failing that, [`.npmignore`](https://docs.npmjs.com/using-npm/developers.html#keeping-files-out-of-your-package).

A more likely dilemma is trying to avoid exposing internal helpers, as per the discussion on [blacklisting specific paths](https://github.com/jkrems/proposal-pkg-exports/issues/46).

Reviewer: Preact recently added `"./": "./"` in exports, and very much exposes scripts and test folders - http://unpkg.com/preact/.

Reviewer: As a rule, in all my tooling that works with `"exports"` I explicitly ignore `"./": "./"` entries due to the level of cruft that creeps in otherwise.

Reviewer 3: I agree we should be clear that disabling `"exports"`'s encapsulation is discouraged.  However I think the cat is out of the bag on the `"./": "./"` capability, and therefore it is best we communicate what it means and why it gets used.  Suggested wording coming up next.

> This will likely be a breaking change. To make the introduction of `"exports"`
> non-breaking, either ensure that every previously supported entry point is
> exported or include an export of the root of the package `"./": "./"`.

Reviewer 3: ```suggestion
This will likely be a breaking change. To make the introduction of `"exports"`
non-breaking, ensure that every previously supported entry point is exported.
It is best to explicitly specify entry points so that the package's public API is
well-defined. As a last resort, package encapsulation can be disabled entirely by
creating an export for the root of the package `"./": "./"` that will expose every file.
```

Reviewer: This seems like a good start. Perhaps an example with `package.json` as well would make sense.

environment, including whether the package is referenced via `require` or via
`import`.

**Warning**: Introducing the `"exports"` field prevents consumers of a package
from using any entry points that are not defined, including the `package.json`.
This will likely be a breaking change. To make the introduction of `"exports"`
non-breaking, ensure that every previously supported entry point is exported.
It is best to explicitly specify entry points so that the package's public API is

> well-defined. As a last resort, package encapsulation can be disabled entirely by

Reviewer: Could we perhaps include an explicit example here? Something like:

> For example, a CommonJS package with `lib/index.js` and `feature/index.js` where users might have previously 
imported either subpath explicitly without explicitly including the `index.js` path, can be written:

```
{
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/index.js": "./lib/index.js",
    "./feature": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

> Since all common existing require patterns are captured in the above, it can be a non-breaking change.

> If you must adopt `"exports"` and really cannot explicitly list subpaths or make a breaking change, then a `"./": "./"` root export can be provided to opt-out of encapsulation.

Perhaps it is overkill, but I just worry a wall of text with one code example like `"./": "./"` will lead to users just copying that directly as the "I just want this to work" path when scanning, unless they can explicitly see what this really involves.
